### PR TITLE
feat(collection methods): add basic scatter method

### PR DESCRIPTION
**Pull request by Paul-Vasylenko**

### CODE REVIEW

File: src/packages/connections/sockets/logic/creators/createSocketCommunicator.ts [hunk 1]
```diff
@@ -113,6 +120,51 @@ export default function createSocketCommunicator({
...
+      throw new Error(
+        `Received data must be an array. Got ${received} from ${process.code}`,
+      );
+    }
+    receiveBuf.push(...received.slice(0, receiveCount));
+
+    return
+  }
+
   function receive(abort?: Abort) {
```
Comment (Ev1ch) [inline src/packages/connections/sockets/logic/creators/createSocketCommunicator.ts:168]: question(blocking): 

there is a conceptual trouble in the current logic. in this method, we just use simple assignments to get value, for example
```ts
const { data } = await communicator.receive();
```
. it means, that we can use `await` only when we can assign value somewhere. 

but in the `scatter` method we can specify the receiving buffer in the method call. so i think we should refactor the `receive` method in the same way.

what do you think?

File: src/packages/connections/sockets/logic/creators/createSocketCommunicator.ts [hunk 2]
```diff
@@ -113,6 +120,51 @@ export default function createSocketCommunicator({
     return send(data, processes, abort);
   }
 
+  async function scatter(
+    data: unknown[],
+    sendCount: number,
+    receiveBuf: unknown[],
+    receiveCount: number,
```
Comment (Ev1ch) [inline src/packages/connections/sockets/logic/creators/createSocketCommunicator.ts:127]: question(blocking): why do we need both `receiveCount` and `sendCount`? tbh, i don't see a case where this numbers can differ

File: src/packages/connections/sockets/logic/creators/createSocketCommunicator.ts [hunk 3]
```diff
@@ -113,6 +120,51 @@ export default function createSocketCommunicator({
...
+      ),
+    );
+
+    const { data: received, process } = await receive(abort);
+    if (!Array.isArray(received)) {
+      throw new Error(
+        `Received data must be an array. Got ${received} from ${process.code}`,
+      );
+    }
+    receiveBuf.push(...received.slice(0, receiveCount));
```
Comment (Ev1ch) [inline src/packages/connections/sockets/logic/creators/createSocketCommunicator.ts:163]: question(blocking): are you sure that we should push into array without cleaning existing items/checking if they aren't present?

**Review by Ev1ch — CHANGES_REQUESTED** [review]

File: src/packages/connections/sockets/logic/creators/createSocketCommunicator.ts [hunk 4]
```diff
@@ -113,6 +120,51 @@ export default function createSocketCommunicator({
...
+      throw new Error('Cant receive more than send');
+    }
+
+    // split data to send
+    const splittedData: Array<Array<unknown>> = [];
+    for (let i = 0; i < allDevicesLength; i++) {
+      const start = i * sendCount;
+      const end = start + sendCount;
+      splittedData.push(data.slice(start, end));
+    }
```
[the remark this answers is not in the document]
Comment (Paul-Vasylenko) [reply src/packages/connections/sockets/logic/creators/createSocketCommunicator.ts:145]: скатер має поділити масив [1,2,3,4,5,6] з параметром 2 на три підмасиви [1,2] [3,4] i [5,6]
Якщо ж в нас були ще якісь елементи, наприклад 7, 8 .... Вони б просто проігнорувалися

Щоб таке пофіксити в майбутньому треба реалізувати ScatterV

File: src/packages/connections/sockets/logic/creators/createSocketCommunicator.ts (hunk 1 above)
Comment (Paul-Vasylenko) [reply src/packages/connections/sockets/logic/creators/createSocketCommunicator.ts:168]: I absolutely agree, that is the problem I faced with buffer may be fixed with changing receive method

File: src/packages/connections/sockets/logic/creators/createSocketCommunicator.ts (hunk 3 above)
Comment (Paul-Vasylenko) [reply src/packages/connections/sockets/logic/creators/createSocketCommunicator.ts:163]: Yes, I think we should clear array at first

**Pull request closed without merging**